libdoipserver/src/DoIPConnection.cpp

```diff
@@ -0,0 +1,219 @@
+    switch(action.type) {
+        case PayloadType::NEGATIVEACK: {
+            //send NACK
+            sendedBytes = sendNegativeAck(action.value);
+
+            if(action.value == _IncorrectPatternFormatCode || 
+                    action.value == _InvalidPayloadLengthCode) {
+                closeSocket();
+                return -1;
+            }
+
+            return sendedBytes;
+        }
+
+        case PayloadType::ROUTINGACTIVATIONREQUEST: {
+            //start routing activation handler with the received message
+            unsigned char result = parseRoutingActivation(data);
+            unsigned char clientAddress [2] = {data[8], data[9]};
+
+            unsigned char* message = createRoutingActivationResponse(logicalGatewayAddress, clientAddress, result);
+            sendedBytes = sendMessage(message, _GenericHeaderLength + _ActivationResponseLength);
+
+            if(result == _UnknownSourceAddressCode || 
+                    result == _UnsupportedRoutingTypeCode) {
+                closeSocket();
+                return -1;
+            } else {
+                //Routing Activation Request was successfull, save address of the client
+                routedClientAddress = new unsigned char[2];
+                routedClientAddress[0] = data[8];
+                routedClientAddress[1] = data[9];
+
+                //start alive check timer
+                if(!aliveCheckTimer.active) {
+                    aliveCheckTimer.cb = std::bind(&DoIPConnection::aliveCheckTimeout,this);
+                    aliveCheckTimer.startTimer();
+                }
+            }
+
+            return sendedBytes;
+        }
+
+        case PayloadType::ALIVECHECKRESPONSE: {
+            return 0;
+        }
+
+        case PayloadType::DIAGNOSTICMESSAGE: {
+
+            unsigned short target_address = 0;
+            target_address |= ((unsigned short)data[10]) << 8U;
+            target_address |= (unsigned short)data[11];
+            bool ack = notify_application(target_address);
+
+            if(ack)
+                parseDiagnosticMessage(diag_callback, routedClientAddress, data, readedBytes - _GenericHeaderLength);
+
+            break;
+        }
+
+        default: {
+            std::cerr << "not handled payload type occured in receiveMessage()" << std::endl;
+            return -1;
+        }
+    }  
+    return -1;
+}
+
+void DoIPConnection::triggerDisconnection() {
+    std::cout << "Application request to disconnect Client from Server" << std::endl;
+    closeSocket();
+}
+
+/**
+ * Sends a message back to the connected client
+ * @param message           contains generic header and payload specific content
+ * @param messageLength     length of the complete message
+ * @return                  number of bytes written is returned,
+ *                          or -1 if error occurred
+ */
+int DoIPConnection::sendMessage(unsigned char* message, int messageLength) {
+    int result = write(tcpSocket, message, messageLength);
```

Comment:
Ich glaub, ich würde die Schreib- und Lese-Zugriffe auf den Socket sicherheitshalber synchronisieren.

Reply:
Das dürfte nicht notwendig sein.
Nach meinem Verständnis sind die socket calls Thread-Safe.

Siehe auch
* https://stackoverflow.com/questions/1981372/are-parallel-calls-to-send-recv-on-the-same-socket-valid
* https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_01

Reply:
Ah, ok. Ich hatte da ein älteres Problem im Kopf, aber da hatten wir mal zwei writes nacheinander, wo das notwendig war. Passt.